packages/experience/src/hooks/use-webauthn-operation.ts

```diff
@@ -0,0 +1,103 @@
+import {
+  MfaFactor,
+  webAuthnRegistrationOptionsGuard,
+  type WebAuthnAuthenticationOptions,
+  type WebAuthnRegistrationOptions,
+} from '@logto/schemas';
+import { trySafe } from '@silverhand/essentials';
+import { startAuthentication, startRegistration } from '@simplewebauthn/browser';
+import type {
+  RegistrationResponseJSON,
+  AuthenticationResponseJSON,
+} from '@simplewebauthn/typescript-types';
+import { useCallback } from 'react';
+import { useTranslation } from 'react-i18next';
+
+import {
+  createWebAuthnRegistrationOptions,
+  generateWebAuthnAuthnOptions,
+} from '@/apis/interaction';
+import { UserMfaFlow } from '@/types';
+
+import useApi from './use-api';
+import useErrorHandler from './use-error-handler';
+import useSendMfaPayload from './use-send-mfa-payload';
+import useToast from './use-toast';
+
+const isRegistrationOptions = (
+  options: WebAuthnRegistrationOptions | WebAuthnAuthenticationOptions
+): options is WebAuthnRegistrationOptions =>
+  webAuthnRegistrationOptionsGuard.safeParse(options).success;
+
+const isAuthenticationResponseJSON = (
+  responseJson: RegistrationResponseJSON | AuthenticationResponseJSON
+): responseJson is AuthenticationResponseJSON => {
+  return 'signature' in responseJson.response;
+};
+
+const useWebAuthnOperation = (flow: UserMfaFlow) => {
+  const { t } = useTranslation();
+  const { setToast } = useToast();
+
+  const asyncCreateRegistrationOptions = useApi(createWebAuthnRegistrationOptions);
+  const asyncGenerateAuthnOptions = useApi(generateWebAuthnAuthnOptions);
+
+  const sendMfaPayload = useSendMfaPayload();
+
+  const handleError = useErrorHandler();
+  const handleRawWebAuthnError = useCallback(
+    (error: unknown) => {
+      if (error instanceof Error) {
+        setToast(error.message);
+        return;
+      }
+
+      setToast(t('error.unknown'));
+    },
+    [setToast, t]
+  );
+
+  return useCallback(async () => {
+    const [error, options] =
+      flow === UserMfaFlow.MfaBinding
+        ? await asyncCreateRegistrationOptions()
+        : await asyncGenerateAuthnOptions();
+
+    if (error) {
+      await handleError(error);
+      return;
+    }
+
+    if (!options) {
+      return;
+    }
+
+    const response = await trySafe(
+      async () =>
+        isRegistrationOptions(options) ? startRegistration(options) : startAuthentication(options),
```

Comment:
Refactor to a `handleWebAuthn` function. Check the type and call relative operation accordingly

Reply:
updated.